src/Middleware/HttpLogging/test/TestW3CLoggerProcessor.cs

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.HttpLogging
+{
+    internal sealed class TestW3CLoggerProcessor : W3CLoggerProcessor
+    {
+        private int _writeCount = 0;
+        private int _expectedWrites;
+        private TaskCompletionSource _tcs;
+        private bool _hasWritten;
+        private readonly object _writeCountLock = new object();
+
+        public TestW3CLoggerProcessor(IOptionsMonitor<W3CLoggerOptions> options, IHostEnvironment environment, ILoggerFactory factory) : base(options, environment, factory)
+        {
+            Lines = new List<string>();
+        }
+
+        public List<string> Lines { get; }
+
+        internal override StreamWriter GetStreamWriter(string fileName)
+        {
+            return StreamWriter.Null;
+        }
+
+        internal override void OnWrite(string message)
+        {
+            Lines.Add(message);
+            lock (_writeCountLock)
+            {
+                _writeCount++;
+                if (_tcs != null && _writeCount >= _expectedWrites)
+                {
+                    _tcs.SetResult();
+                }
+            }
+        }
+
+        public Task WaitForWrites(int numWrites)
+        {
+            lock (_writeCountLock)
+            {
+                if (_writeCount >= numWrites)
+                {
+                    return Task.CompletedTask;
+                }
+                _expectedWrites = numWrites;
+                _tcs = new TaskCompletionSource();
```

Comment:
nit: We generally pass `TaskCreationOptions.RunContinuationsAsynchronously` to TCS's unless there is a very good reason not to.

Reply:
FYI By not doing this, that means the end of the test method can run *inside* the call to `_tcs.SetResult()`. This can create non-deterministic deadlocks sometimes if the test tries to do a blocking dispose of the host without anything else truly async in-between the call to WaitForWrites and disposing the host. Since you're using `await using`, that shouldn't be a problem here, but there's also no reason not to.